extruct/jsonld.py

```diff
@@ -16,26 +16,24 @@
 class JsonLdExtractor(object):
     _xp_jsonld = lxml.etree.XPath('descendant-or-self::script[@type="application/ld+json"]')
 
-    def extract(self, htmlstring, base_url=None, encoding="UTF-8"):
+    def extract(self, htmlstring, base_url=None, encoding="UTF-8", as_json=False):
         tree = parse_html(htmlstring, encoding=encoding)
-        return self.extract_items(tree, base_url=base_url)
+        return self.extract_items(tree, base_url=base_url, as_json=as_json)
 
-    def extract_items(self, document, base_url=None):
+    def extract_items(self, document, base_url=None, as_json=False):
         return [
             item
-            for items in map(self._extract_items, self._xp_jsonld(document))
+            for items in map(self._extract_items_raw if as_json else self._extract_items, self._xp_jsonld(document))
             if items for item in items if item
         ]
 
+    def _extract_items_raw(self, node):
+        return HTML_OR_JS_COMMENTLINE.sub('', node.xpath('string()'))
+
     def _extract_items(self, node):
-        script = node.xpath('string()')
-        try:
-            # TODO: `strict=False` can be configurable if needed
-            data = json.loads(script, strict=False)
-        except ValueError:
-            # sometimes JSON-decoding errors are due to leading HTML or JavaScript comments
-            data = json.loads(
-                HTML_OR_JS_COMMENTLINE.sub('', script), strict=False)
+        script = self._extract_items_raw(node)
+        # TODO: `strict=False` can be configurable if needed
+        data = json.loads(script, strict=False)
```

Comment:
Why is try/except removed here?

Reply:
Original code tries to load json and then remove comments. I can't imagine scenario where someone would like to receive json with javascript/html comments in tackt. I took the liberty of simplifying the process and explicitly remove the comments as it's relatively cheap operation. If you disagree we can change it back to "trying to remove comments if loading fails" approach.

Reply:
@Granitosaurus ah I didn't notice that it's still there but was moved, thanks for pointing it out. I'd rather not try removing comments by default, because they can be in valid JSON:
```
>>> json.dumps(['<!-- comment -->'])
'["<!-- comment -->"]'
```
so removing comments it a heuristics which is not always correct. So I think it would be better to do just `node.xpath('string()'))` in `_extract_items_raw`, and try removing comments only if json parsing fails, as in master.

Reply:
Sure thing, reverted!